main.tf

```diff
@@ -1,8 +1,47 @@
 locals {
+  # Define resource types that support cold storage
+  supported_resource_types = [
+    "arn:aws:cloudformation:*:*:stack/*",
+    "arn:aws:dynamodb:*:*:table/*",
+    "arn:aws:ec2:*:*:volume/*",
+    "arn:aws:elasticfilesystem:*:*:file-system/*",
+    "arn:aws:ec2:*:*:instance/*", # For SAP HANA on EC2
+    "arn:aws:timestream:*:*:database/*",
+    "arn:aws:backup:*:*:backup-vault/*",
+    "arn:aws:backup:*:*:recovery-point/*"
+  ]
+
+  # Determine the SNS topic ARN to use
   sns_topic_arn = var.sns_topic_arn == "" ? aws_sns_topic.bkup_sns_topic[0].arn : var.sns_topic_arn
+
+  # Identify resources that don't support cold storage
+  cold_storage_unsupported_resources = [
+    for arn in var.source_arns : arn
+    if !can(regex(join("|", local.supported_resource_types), arn))
+  ]
+
+  # Error message for resources that don't support cold storage
+  cold_storage_error_message = "Error: Cold storage is not supported for the following resources: ${join(", ", local.cold_storage_unsupported_resources)}."
+}
+
+# Validation resource to check if cold storage is enabled for unsupported resources
+resource "null_resource" "cold_storage_validation" {
+  count = var.cold_storage_after != null && length(local.cold_storage_unsupported_resources) > 0 ? 1 : 0
+
+  provisioner "local-exec" {
+    command = "echo '${local.cold_storage_error_message}' && exit 1"
+  }
 }
 
-# Encryption key for the Backup Vault
+# Validation to ensure delete_after is at least 90 days more than cold_storage_after
+resource "null_resource" "lifecycle_validation" {
+  count = var.cold_storage_after != null && var.delete_after != null && (var.delete_after - var.cold_storage_after) < 90 ? 1 : 0
+
+  provisioner "local-exec" {
+    command = "echo 'Error: delete_after must be at least 90 days more than cold_storage_after' && exit 1"
+  }
+}
```

Comment:
Seems like this could be done in a `check` block. Something like this, perhaps:
```hcl
check "health_check" {
  assert {
    condition     = var.cold_storage_after == null || var.delete_after == null || (var.delete_after - var.cold_storage_after) >= 90
    error_message = "Error: delete_after must be at least 90 days more than cold_storage_after"
  }
}
```

Reply:
Updated the validation block using the check